File: tools/hollow-knight-patches/src/skins/core/SkinSessionModels.cs
namespace DualSouls.Skins.HollowKnight.Core
{
    public enum SessionLeasePhase { LAUNCH_PENDING, GAME_OWNED, CLOSED }
    public enum SessionMutationGate { UNKNOWN, ACTIVE, CLEAR }
    public enum SessionClaimUse { AVAILABLE, IN_FLIGHT, CONSUMED, REJECTED }
    public enum SessionOperation { CLAIM, CLOSE }
    public enum SessionLiveness { UNKNOWN, ALIVE, DEAD }
    public enum SessionPresence { UNKNOWN, PRESENT, ABSENT }
    public enum SessionFailure { DURABILITY_UNAVAILABLE, INDETERMINATE, LIFECYCLE_BLOCKED, SESSION_RECOVERY_AMBIGUOUS, PROFILE_QUOTA_EXCEEDED, DOCUMENT_INVALID }
    public sealed record SessionProcessIdentity(int Uid, int Pid, string ProcessStartToken);
    // TokenSha256 is the digest of decoded token bytes, never an executable credential.
    public sealed record SessionBinding(string ProfileId, string DescriptorId, string DescriptorSha256, string DescriptorPath,
        string LeaseId, string TokenSha256, long SessionSequence, string RegistryGenerationId, string RegistrySha256, SessionProcessIdentity LauncherOwner);
    public sealed record SessionLeaseHead(string DescriptorId, string LeaseId, long TransitionSequence, SessionLeasePhase State, string Sha256);
    public sealed record SessionLeaseDocument(int SchemaVersion, SessionBinding Binding, long TransitionSequence, string TransitionId,
        string ParentTransitionId, SessionLeasePhase State, SessionProcessIdentity GameOwner, string CloseReason);
    // Trusted locked-store projection: canonical bytes were verified against this digest.
    // Construction establishes no runtime truth. No hashing, parsing, or unlocked enrichment here.
    public sealed record VerifiedSessionLease(SessionLeaseHead Head, SessionLeaseDocument Document, string CanonicalDocumentSha256);
    public sealed record SessionCorrelation(SessionBinding Binding, long Ordinal);
    public abstract record SessionBarrier
    {
        private SessionBarrier() { }
        public sealed record ActiveInstalled(SessionLeaseHead Head) : SessionBarrier;
        public sealed record ActiveRemoved(SessionLeaseHead ExpectedParent) : SessionBarrier;
    }
    // Coordinator attests completed descriptor qualification, pending publication and global active barrier.
    public sealed record SessionAcquisitionEvidence(VerifiedSessionLease Pending, SessionBarrier.ActiveInstalled Barrier);
    public sealed record SessionTransitionReceipt(SessionCorrelation Correlation, SessionOperation Operation, VerifiedSessionLease Parent,
        VerifiedSessionLease Child, SessionBarrier Barrier);
    public abstract record SessionTransitionResult
    {
        private SessionTransitionResult() { }
        public sealed record Durable(SessionTransitionReceipt Receipt) : SessionTransitionResult;
        public sealed record Failed(SessionFailure Code) : SessionTransitionResult;
    }
    public sealed record SessionTargetProcess(string PackageName, string ProcessName);
    public sealed record SessionOwnerEvidence(SessionCorrelation Correlation, SessionBinding Binding, SessionProcessIdentity ExpectedOwner,
        SessionLiveness Liveness, SessionProcessIdentity AliveOwner);
    // Trusted verifier must use the exact build-configured target for this binding and operation.
    // Missing configuration, denied queries and PID/start mismatches are UNKNOWN, never DEAD/ABSENT.
    public sealed record SessionTargetEvidence(SessionCorrelation Correlation, SessionBinding Binding, SessionTargetProcess Target,
        SessionPresence Presence, SessionProcessIdentity PresentOwner);
    // Attests COMPLETE bounded profile validation/reconciliation and no remaining open lease.
    // Denied/malformed/ambiguous scans cannot produce this evidence. Constructor is not a verifier.
    public sealed record SessionProfileRecoveryEvidence(string ProfileId, SessionCorrelation Correlation);
    public abstract record SessionRecoveryResult
    {
        private SessionRecoveryResult() { }
        public sealed record NoOpenLease(SessionProfileRecoveryEvidence Profile) : SessionRecoveryResult;
        public sealed record RetainedOpenLease(VerifiedSessionLease Lease, SessionBarrier.ActiveInstalled Barrier, SessionOwnerEvidence Owner,
            SessionTargetEvidence Target, SessionTransitionReceipt ObservedClaim = null) : SessionRecoveryResult;
        public sealed record RecoveredClosed(VerifiedSessionLease Parent, SessionTransitionReceipt Receipt, SessionOwnerEvidence Owner,
            SessionTargetEvidence Target, SessionProfileRecoveryEvidence Profile) : SessionRecoveryResult;
        public sealed record Deferred(SessionFailure Code) : SessionRecoveryResult;
    }
    public sealed record SessionRecoveryRecord(SessionCorrelation Correlation, SessionRecoveryResult Result);
    public abstract record SessionCommand(SessionCorrelation Correlation)
    {
        public sealed record ClaimExistingBinding(SessionCorrelation Correlation, SessionBinding Binding, SessionProcessIdentity Owner) : SessionCommand(Correlation);
        public sealed record CloseExistingBinding(SessionCorrelation Correlation, SessionBinding Binding, string Reason) : SessionCommand(Correlation);
        public sealed record RecoverExistingSessionStore(SessionCorrelation Correlation, SessionBinding ExpectedBinding) : SessionCommand(Correlation);
    }
    public abstract record SessionEvent
    {
        private SessionEvent() { }
        public sealed record BindIssuedPending(SessionAcquisitionEvidence Evidence) : SessionEvent;
        public sealed record ClaimRequested(SessionBinding Binding, SessionProcessIdentity Owner) : SessionEvent;
        public sealed record ClaimCompleted(SessionCorrelation Correlation, SessionTransitionResult Result) : SessionEvent;
        public sealed record CloseRequested(SessionBinding Binding, string Reason) : SessionEvent;
        public sealed record CloseCompleted(SessionCorrelation Correlation, SessionTransitionResult Result) : SessionEvent;
        public sealed record RecoveryRequested : SessionEvent;
        public sealed record RecoveryCompleted(SessionCorrelation Correlation, SessionRecoveryResult Result) : SessionEvent;
    }
    // Pending recovery preserves exact-owner retry eligibility. Observed owned recovery is
    // close/recovery-only (REJECTED), never a successful bridge claim. A consumed claim stays
    // consumed after recovery but loses write eligibility; a fresh core is integration-owned.
    // One bounded observation retained even when later recovery is deferred. Equal ordinals
    // identify recovery-observed ownership, not a successful bridge claim; later recovery
    // revokes a previously consumed claim. Neither shape is runtime proof by construction.
    public sealed record SessionClaimRecoveryFence(SessionCorrelation RecoveryCorrelation, SessionTransitionReceipt ClaimReceipt);
    public sealed record SessionState(SessionAcquisitionEvidence Acquisition = null, VerifiedSessionLease Lease = null,
        SessionClaimUse ClaimUse = SessionClaimUse.AVAILABLE, SessionProcessIdentity PinnedOwner = null, string PinnedCloseReason = null,
        SessionCommand Pending = null, SessionMutationGate Gate = SessionMutationGate.UNKNOWN, SessionTransitionReceipt ClaimReceipt = null,
        SessionTransitionReceipt ClosedReceipt = null, SessionRecoveryRecord Recovery = null, long RecoveryHighWater = 0, long Ordinal = 0,
        SessionCommand.ClaimExistingBinding ClaimAttempt = null, SessionClaimRecoveryFence ClaimRecoveryFence = null,
        // Last explicit request: retry ordinals advance, but binding/reason stay pinned.
        // Recovery preserves it and retains its independently qualified terminal audit;
        // that audit survives later profile recovery and is never an explicit-close success.
        SessionCommand.CloseExistingBinding CloseAttempt = null, SessionRecoveryRecord RecoveryClosure = null);
    // At most one immutable command; no mutable collection or captured I/O capability.
    public sealed record SessionDecision(SessionState State, string Diagnosis, SessionCommand Command = null, SessionTransitionReceipt CompletedClose = null);
}

File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinSessionModels.kt
package dev.silksong.launcher.skins.core

enum class SessionLeasePhase { LAUNCH_PENDING, GAME_OWNED, CLOSED }
enum class SessionMutationGate { UNKNOWN, ACTIVE, CLEAR }
enum class SessionClaimUse { AVAILABLE, IN_FLIGHT, CONSUMED, REJECTED }
enum class SessionOperation { CLAIM, CLOSE }
enum class SessionLiveness { UNKNOWN, ALIVE, DEAD }
enum class SessionPresence { UNKNOWN, PRESENT, ABSENT }
enum class SessionFailure { DURABILITY_UNAVAILABLE, INDETERMINATE, LIFECYCLE_BLOCKED, SESSION_RECOVERY_AMBIGUOUS, PROFILE_QUOTA_EXCEEDED, DOCUMENT_INVALID }
data class SessionProcessIdentity(val uid: Int, val pid: Int, val processStartToken: String)
/** TokenSha256 is the digest of decoded token bytes, never an executable credential. */
data class SessionBinding(val profileId: String, val descriptorId: String, val descriptorSha256: String, val descriptorPath: String,
    val leaseId: String, val tokenSha256: String, val sessionSequence: Long, val registryGenerationId: String, val registrySha256: String, val launcherOwner: SessionProcessIdentity)
data class SessionLeaseHead(val descriptorId: String, val leaseId: String, val transitionSequence: Long, val state: SessionLeasePhase, val sha256: String)
data class SessionLeaseDocument(val schemaVersion: Int, val binding: SessionBinding, val transitionSequence: Long, val transitionId: String,
    val parentTransitionId: String?, val state: SessionLeasePhase, val gameOwner: SessionProcessIdentity?, val closeReason: String?)
/** Trusted locked-store projection: canonical bytes verified against this digest. Construction proves no runtime truth. */
data class VerifiedSessionLease(val head: SessionLeaseHead, val document: SessionLeaseDocument, val canonicalDocumentSha256: String)
data class SessionCorrelation(val binding: SessionBinding?, val ordinal: Long)
sealed interface SessionBarrier {
    data class ActiveInstalled(val head: SessionLeaseHead) : SessionBarrier
    data class ActiveRemoved(val expectedParent: SessionLeaseHead) : SessionBarrier
}
/** Coordinator attests descriptor qualification, pending publication and the global active barrier. */
data class SessionAcquisitionEvidence(val pending: VerifiedSessionLease, val barrier: SessionBarrier.ActiveInstalled?)
data class SessionTransitionReceipt(val correlation: SessionCorrelation, val operation: SessionOperation, val parent: VerifiedSessionLease,
    val child: VerifiedSessionLease, val barrier: SessionBarrier?)
sealed interface SessionTransitionResult {
    data class Durable(val receipt: SessionTransitionReceipt?) : SessionTransitionResult
    data class Failed(val code: SessionFailure) : SessionTransitionResult
}
data class SessionTargetProcess(val packageName: String, val processName: String)
data class SessionOwnerEvidence(val correlation: SessionCorrelation, val binding: SessionBinding, val expectedOwner: SessionProcessIdentity,
    val liveness: SessionLiveness, val aliveOwner: SessionProcessIdentity?)
/** Verifier uses the exact build-configured target; missing configuration, denied queries and PID/start mismatches are UNKNOWN. */
data class SessionTargetEvidence(val correlation: SessionCorrelation, val binding: SessionBinding, val target: SessionTargetProcess,
    val presence: SessionPresence, val presentOwner: SessionProcessIdentity?)
/** COMPLETE bounded profile validation/reconciliation with no open lease. Constructor is not a verifier. */
data class SessionProfileRecoveryEvidence(val profileId: String, val correlation: SessionCorrelation)
sealed interface SessionRecoveryResult {
    data class NoOpenLease(val profile: SessionProfileRecoveryEvidence?) : SessionRecoveryResult
    data class RetainedOpenLease(val lease: VerifiedSessionLease, val barrier: SessionBarrier.ActiveInstalled?, val owner: SessionOwnerEvidence?,
        val target: SessionTargetEvidence?, val observedClaim: SessionTransitionReceipt? = null) : SessionRecoveryResult
    data class RecoveredClosed(val parent: VerifiedSessionLease, val receipt: SessionTransitionReceipt?, val owner: SessionOwnerEvidence?,
        val target: SessionTargetEvidence?, val profile: SessionProfileRecoveryEvidence?) : SessionRecoveryResult
    data class Deferred(val code: SessionFailure) : SessionRecoveryResult
}
data class SessionRecoveryRecord(val correlation: SessionCorrelation, val result: SessionRecoveryResult)
sealed interface SessionCommand {
    val correlation: SessionCorrelation
    data class ClaimExistingBinding(override val correlation: SessionCorrelation, val binding: SessionBinding, val owner: SessionProcessIdentity) : SessionCommand
    data class CloseExistingBinding(override val correlation: SessionCorrelation, val binding: SessionBinding, val reason: String) : SessionCommand
    data class RecoverExistingSessionStore(override val correlation: SessionCorrelation, val expectedBinding: SessionBinding?) : SessionCommand
}
sealed interface SessionEvent {
    data class BindIssuedPending(val evidence: SessionAcquisitionEvidence?) : SessionEvent
    data class ClaimRequested(val binding: SessionBinding?, val owner: SessionProcessIdentity?) : SessionEvent
    data class ClaimCompleted(val correlation: SessionCorrelation?, val result: SessionTransitionResult?) : SessionEvent
    data class CloseRequested(val binding: SessionBinding?, val reason: String?) : SessionEvent
    data class CloseCompleted(val correlation: SessionCorrelation?, val result: SessionTransitionResult?) : SessionEvent
    data object RecoveryRequested : SessionEvent
    data class RecoveryCompleted(val correlation: SessionCorrelation?, val result: SessionRecoveryResult?) : SessionEvent
}
/** Pending recovery preserves exact-owner retry. Observed owned recovery is close/recovery-only (REJECTED),
 * never a successful bridge claim. Consumed claims stay consumed but lose writes after recovery.
 * A fresh core belongs to integration, not a reset event on this state. */
/** One bounded observation survives later deferred recovery. Equal ordinals mean recovery-observed
 * ownership, not a successful bridge claim; later recovery revokes a consumed claim.
 * Neither shape establishes runtime truth by construction. */
data class SessionClaimRecoveryFence(val recoveryCorrelation: SessionCorrelation, val claimReceipt: SessionTransitionReceipt)
data class SessionState(val acquisition: SessionAcquisitionEvidence? = null, val lease: VerifiedSessionLease? = null,
    val claimUse: SessionClaimUse = SessionClaimUse.AVAILABLE, val pinnedOwner: SessionProcessIdentity? = null, val pinnedCloseReason: String? = null,
    val pending: SessionCommand? = null, val gate: SessionMutationGate = SessionMutationGate.UNKNOWN, val claimReceipt: SessionTransitionReceipt? = null,
    val closedReceipt: SessionTransitionReceipt? = null, val recovery: SessionRecoveryRecord? = null, val recoveryHighWater: Long = 0, val ordinal: Long = 0,
    val claimAttempt: SessionCommand.ClaimExistingBinding? = null, val claimRecoveryFence: SessionClaimRecoveryFence? = null,
    // Last explicit request: retry ordinals advance, binding/reason stay pinned.
    // Recovery preserves it and retains its independently qualified terminal audit;
    // that audit survives later profile recovery and is never an explicit-close success.
    val closeAttempt: SessionCommand.CloseExistingBinding? = null, val recoveryClosure: SessionRecoveryRecord? = null)
/** At most one immutable command; no mutable collection or captured I/O capability. */
data class SessionDecision(val state: SessionState?, val diagnosis: String, val command: SessionCommand? = null, val completedClose: SessionTransitionReceipt? = null)
